### ISSUE TITLE
Union of tuples differing by a single member is not considered equivalent to a tuple containing a single union

### ISSUE THREAD

**Issue by Starwort** [microsoft/pyright#8068]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**
In invariant position, a union of tuples is not considered equivalent to the (equivalent) union-containing tuple.

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
from typing import Literal

some_data = [1, 2, 3, 4, 5]
some_other_data = [2, 3, 4, 5, 6]


def get_best_data() -> list[tuple[Literal["sourceA", "sourceB"], int]]:
    return sorted(
        [
            *(("sourceA", i) for i in some_data),
            *(("sourceB", i) for i in some_other_data),
        ],
        key=lambda i: i[1],
    )[:5]


test_a: list[Literal["sourceA"] | Literal["sourceB"]] = [
    i[0] for i in get_best_data()
]  # works fine when not wrapped in tuple

test_b: list[tuple[Literal["sourceA"]] | tuple[Literal["sourceB"]]] = [
    i[:1] for i in get_best_data()
]  # breaks when wrapped in tuple
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

- pyright 1.1.336 (command-line)
- Pylance v2024.5.1 (in VSCode)

**Comment by erictraut** [comment]
I recently wrote a new chapter for the official Python typing spec about tuples. This chapter discusses [subtyping rules](https://typing.readthedocs.io/en/latest/spec/tuples.html#type-compatibility-rules) for tuples and mentions that type checkers _may_ use tuple expansion (and by implication, tuple fusion). We decided not to specify all of the places where type checkers _must_ use tuple expansion or fusion because these operations are complex and expensive to perform at analysis time.

None of the major Python type checkers currently perform tuple fusion as you have suggested in your bug report. Here's a simplified code sample that demonstrates current behaviors.

```python
def func1(a: tuple[int] | tuple[str]):
    b: tuple[int | str] = a  # No error

def func2(a: tuple[int | str]):
    b: list[tuple[int] | tuple[str]] = a  # Type error from pyright and mypy

def func3(a: list[tuple[int] | tuple[str]]):
    b: list[tuple[int | str]] = a  # Type error from pyright and mypy
```

I've changed this issue from a bug report to an enhancement request. I'm going to reject the enhancement request for now, but I am open to reconsidering it in the future if there is sufficient signal from other pyright users that it would be useful, e.g. in the form of thumbs-up reactions in this issue.
